server/drivers/yard2LCD.c

```diff
@@ -0,0 +1,647 @@
+*/
+	tmp = DEFAULT_OFF_BRIGHTNESS;
+	p->off_brightness = tmp;
+
+
+	report(RPT_DEBUG, "%s: Init done", drvthis->name);
+	return 0;
+}
+
+/*
+ * Closes down the driver
+ */
+MODULE_EXPORT void
+yard_close(Driver *drvthis)
+{
+	debug(RPT_DEBUG, "%s: Event 09 - Enter yard_close",drvthis->name);
+	PrivateData *p = (PrivateData *)drvthis->private_data;
+	
+	if (p != NULL) {
+		close(p->fd);
+		if (p->framebuf)
+			free(p->framebuf);
+		free(p);
+	}
+
+	drvthis->store_private_ptr(drvthis, NULL);
+}
+
+/*
+ * Returns the display width in characters
+ */
+MODULE_EXPORT int
+yard_width(Driver *drvthis)
+{
+	debug(RPT_DEBUG, "%s: Event 10 - Enter yard_width",drvthis->name);
+	PrivateData *p = (PrivateData *)drvthis->private_data;
+
+	return p->width;
+}
+
+/*
+ * Returns the display height in characters
+ */
+MODULE_EXPORT int
+yard_height(Driver *drvthis)
+{
+	debug(RPT_DEBUG, "%s: Event 11 - Enter yard_height",drvthis->name);
+	PrivateData *p = (PrivateData *)drvthis->private_data;
+
+	return p->height;
+}
+
+/*
+ * Clears the framebuffer
+ */
+MODULE_EXPORT void
+yard_clear(Driver *drvthis)
+{
+	debug(RPT_DEBUG, "%s: Event 12 - Enter yard_clear",drvthis->name);
+	PrivateData *p = (PrivateData *)drvthis->private_data;
+	
+	// Reset custom char mode and clear frambuffer
+	p->ccmode = standard;
+	memset(p->framebuf, ' ', p->width * p->height);
+}
+
+/*
+ * Flushes a single frame to the VFD
+ */
+MODULE_EXPORT void
+yard_flush(Driver *drvthis)
+{
+	debug(RPT_DEBUG, "%s: Event 13 - Enter yard_flush",drvthis->name);
+	PrivateData *p = (PrivateData *)drvthis->private_data;
+	int i;
+	int width, height;
+
+	switch (p->dispmode) {
+		case Dmode_standard: 
+		{
+			// Set width and height
+			width = p->width;
+			height = p->height;
+			break;
+		}
+		default: 
+		{
+			debug(RPT_DEBUG, "%s: Event 24 - Only Char. Mode supported: %d",drvthis->name,p->dispmode);
+			return;
+			break;
+		}
+	}
+	
+	// Send framebuffer content to LCD
+	for (i = 0; i < height; i++) 
+	{
+		// Line feed
+		yard_hwGotoXY(drvthis, 1, i+1);
+			
+		// Write single line
+		yard_hwPrintCharArray(drvthis, &(p->framebuf[i*width]), width);
+		debug(RPT_DEBUG, "YARD2: flushed line %d - chars %d", i,width);
+	}
+
+
+}
+
+/*
+ * Prints a string on the VFD
+ */
+MODULE_EXPORT void
+yard_string(Driver *drvthis, int x, int y, const char string[])
+{
+	debug(RPT_DEBUG, "%s: Event 14 - Enter yard_string: X.%d - Y.%d - %s",drvthis->name,x,y,string);
+
+	PrivateData *p = (PrivateData *)drvthis->private_data;
+	int i;
+
+	// Change display mode if necessary; Only char mode 0 is supported right now
+	if (p->dispmode == Dmode_graphical) 
+	{
+		yard_hwClearLCD(drvthis);
+		p->dispmode = Dmode_standard;
+	}
+	
+	x--; //start with 0,0
+	y--;
+	for (i = 0; string[i] != '\0'; i++) 
+	{
+		// Check for buffer overflow
+		if ( (y * p->width + x + i) >= (p->width * p->height) )
+			break;
+		p->framebuf[(y * p->width) + x + i] = string[i];
+	}
+}
```

Comment:
This function should check if the string actually fits on the line. See:
http://lcdproc.sourceforge.net/docs/current-dev.html#function-details